hilti/toolchain/include/ast/node.h

```diff
@@ -464,6 +464,30 @@ class Node {
         return n;
     }
 
+    /**
+     * Returns true if the node's children contain a particular child node.
+     *
+     * @param n child node to look for
+     * @param recurse if true, will also check children of children recursively
+     */
+    bool hasChild(const Node* n, bool recurse = false) const {
+        if ( ! n )
+            return false;
+
+        if ( std::find(_children.begin(), _children.end(), n) != _children.end() )
+            return true;
+
+        if ( ! recurse )
+            return false;
+
+        for ( const auto* c : _children ) {
+            if ( c && c->hasChild(n, true) )
+                return true;
+        }
+
+        return false;
```

Comment:
Nit: I'd prefer passing `recurse` instead of introducing the dependency on the early return just above. We could also write more clearly as

```suggestion
        return std::any_of(_children.begin(), _children.end(),
                           [&](const auto* c) { return c && c->hasChild(n, recurse); });
```